Stats Traffic: Adapt existing views for a new Traffic tab

staskus opened the pull request:
Fixes #22443

Adapt existing Period views for a new Traffic tab and its styling.

Based on the conclusions from Analyze High-Level Stats architecture (https://github.com/wordpress-mobile/WordPress-iOS/issues/22383):

1. Created deprecated duplicates of `SiteStatsPeriodTableView` to be used for Days/Months/Weeks/Years tabs, so we could more easily modify `SiteStatsPeriodTableView` for Traffic tab without complicating development
2. Started using the same base class for Insights and Traffic tabs to share table view configuration and styles
3. Set `insetGrouped` style to show cards
4. Started displaying rows within `SiteStatsPeriodVC` in separate sections
5. Used `StatsBaseClass` mechanism to display titles within cell rather than in a separate row
6. Adjusted spacing

<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/336cfd97-1494-48bc-9a37-83f957770eb2" width="200">

## To test:

### Stats Traffic Feature Flag Disabled

Days/Weeks/Months/Years tabs should work as before

### Stats Traffic Feature Flag enabled

- Traffic tab should display same rows as Days/Weeks/Months/Years tabs but with a new card style (except for bar chart and date picker which will be replaced later)
- Traffic tab and Insights tab should have common spacing and look for table view

## Regression Notes
1. Potential unintended areas of impact

Breaking existing Days/Weeks/Months/Years tabs. Part of my conclusion from the analysis is that we'll want to reuse the majority of the code but also make improvements and changes. 

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Creating new deprecated classes should help avoid all of the risks.

3. What automated tests I added (or what prevented me from doing so)

None so far

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

staskus committed:
Create SiteStatsPeriodTableViewControllerDeprecated

SiteStatsPeriodTableViewControllerDeprecated is a duplicated version of SiteStatsPeriodTableViewController to be used while Stats Traffic feature flag is disabled. It's created with the purpose to unblock Stats Traffic development and additions to SiteStatsPeriodTableViewController without having to support 2 configurations on the same file at once.

staskus committed:
Create SiteStatsPeriodViewModelDeprecated

SiteStatsPeriodViewModelDeprecated is a duplicated version of SiteStatsPeriodViewModel to be used while Stats Traffic feature flag is disabled. It's created with the purpose to unblock Stats Traffic development and make additions to SiteStatsPeriodViewModel without having to support 2 configurations on the same file at once.

staskus committed:
Load SiteStatsPeriodViewController when Traffic tab is selected

staskus committed:
Remove SiteStatsPeriodTableViewController from Storyboard

Storyboard just contains an empty view controller, it's not needed

staskus committed:
Make PeriodVC to be SiteStatsPeriodTableViewController with insetGrouped style

staskus committed:
Put each row into a separate section to support card layout

staskus committed:
Add title to postsAndPages row

staskus committed:
Add title to referrers row

staskus committed:
Add title to clicks row

staskus committed:
Add title to authors row

staskus committed:
Manually build sections

staskus committed:
Add title to countries row

staskus committed:
Add title to search terms row

staskus committed:
Add title to published row

staskus committed:
Add title to video row

staskus committed:
Add title to downloads row

staskus committed:
Added TODOs for new elements

staskus committed:
Update section spacing for Traffic and Spacing

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Remove section headers from loading and empty states

staskus committed:
Merge branch 'trunk' into task/22443-stats-traffic-adapt-existing-views-for-a-new-traffic-tab

staskus committed:
Turn ghostable loading cell into a card with visible header

Turn all ghostable cells StatsBaseCell but set topConstraint as optional so subclasses that don't implement topConstraint continue to work working as before

staskus commented:
Added one more commit to make loading state of the cell appear as a card and with a header

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/69a6ae79-1fa7-4524-8bec-a8e68470e074

guarani commented on WordPress/Classes/ViewRelated/Stats/Insights/StatsBaseCell.swift:
```diff
@@ -43,20 +43,20 @@ class StatsBaseCell: UITableViewCell {
     }()
 
-    @IBOutlet var topConstraint: NSLayoutConstraint!
+    @IBOutlet var topConstraint: NSLayoutConstraint?
```
nitpick: Is this ever `nil`? All the cells that use `StatsBaseCell` have XIBs that set `topConstraint`, but I feel I must be missing something since it was changed here. I did a quick test by changing this back to an implicitly unwrapped optional and no crashes (tested with the feature flag on and off).

staskus replied:
There are ghost cells that do not have this topConstraint implemented

<img width="230" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/96a41940-4fd8-4616-b2b7-32738d0fd658">

I made `StatsGhostBaseCell` to be `StatsBaseCell` and set `topConstraint` for `StatsTopCell` but not for others. Most of them are used within Insights view, and not Traffic view. I'll check maybe it's not too much of an effort to set it for the other cells as well.

staskus replied:
I made all these cells have `topConstraint` and most of these ghost cells to set `statSection` (and consequentially header title). Some of the cells are more used in generic loading cases so they don't need `statSection` to be set.

I think now Insight view loading looks more tidy and will be consistent with Traffic view:

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/5013666b-6603-4a81-b4d5-4c978e5f3088

guarani replied:
I see! I forgot to check the `StatsBaseCell` subclasses. The loading states with titles look good ✨

guarani reviewed (approved):
Nice work @staskus! I tested the old and new flows and things are looking good. It's good to see a lot of the old code being reusable.

staskus committed:
Support heading in StatsGhostGrowAudienceCell by setting topConstraint

staskus committed:
Support heading in StatsGhostPostingActivityCell by setting topConstraint

staskus committed:
Support heading in StatsGhostTabbedCell by setting topConstraint

staskus committed:
Set top constraint for StatsGhostTopHeaderCell

staskus committed:
Support heading in StatsGhostTwoColumnCell by setting topConstraint

staskus committed:
Support heading in StatsGhostChartCell by setting topConstraint

staskus committed:
Set top constraint for StatsGhostSingleRowCell

staskus committed:
Set top constraint for StatsGhostTitleCell

staskus committed:
Set title for remaining ghost rows in Stats Insights
